feat(visible): Add visible() method on Wrapper (#327)

Toilal opened the pull request:
Ready ! :)

eddyerburgh commented:
Thanks for the PR 😀

I don't think we should add a parent method as part of this PR, or at all —https://github.com/vuejs/vue-test-utils/issues/65

You should use the element parentNode/parentElement properties to traverse the parent tree. 

Like you found, most vnodes have an undefined parent. Functional vnodes use the parent property.

Toilal commented:
I just removed `parent()`, implement `visible()` using HTMLElement, and added tests. It seems to work now :p. 

`visible()` is also consistent with `exists()` (if `exists() == false` then `visible() == false`)

eddyerburgh commented on package.json:
```diff
@@ -10,7 +10,7 @@
   "scripts": {
     "build": "node build/build.js",
-    "build:test": "NODE_ENV=test node build/build.js",
+    "build:test": "cross-env NODE_ENV=test node build/build.js",
```
Thanks, good spot

eddyerburgh commented on src/wrappers/wrapper.js:
```diff
@@ -120,6 +120,30 @@ export default class Wrapper implements BaseWrapper {
+   */
+  visible (): boolean {
+    if (!this.exists()) {
+      console.log('NOT EXISTS')
```
We should throw an error here using the `throwError` function:
`throwError('cannot call visible on a wrapper without an element')`

eddyerburgh commented on src/wrappers/wrapper.js:
```diff
@@ -120,6 +120,30 @@ export default class Wrapper implements BaseWrapper {
+   * Utility to check wrapper is visible. Returns false if a parent element has display: none or visibility: hidden style.
+   */
+  visible (): boolean {
+    if (!this.exists()) {
```
You should check for an element here instead of exists—`if(!this.element)`. That's how we check for an element in other tests (I think this might solve the flow error)

eddyerburgh commented on src/wrappers/wrapper.js:
```diff
@@ -120,6 +120,30 @@ export default class Wrapper implements BaseWrapper {
+      }
+      element = element.parentElement
+    }
+    console.log('no parent')
```
We should remove these console logs

eddyerburgh reviewed (changes requested):
Looks good, just a couple of changes. Also we need the flow error to be fixed

eddyerburgh commented:
Are you able to make these changes @Toilal ?

Toilal commented:
Oops, I just forget this one, i'll do ASAP

Toilal commented:
Done

38elements commented:
I think it is necessary to add this to the document.

eddyerburgh commented:
Yes this should be added to the documentation before merging

On 14 Jan 2018 14:21, "38elements" <notifications@github.com> wrote:

> I think it is necessary to add this to the document.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/vuejs/vue-test-utils/pull/335#issuecomment-357515068>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMlbW-sJq-X7_hmFySwR4CiB9VkiFl1Lks5tKg1XgaJpZM4RWjoh>
> .
>

eddyerburgh commented:
@Toilal Can you add a page on visible to docs/en/wrapper/api/visible.md, docs/en/api/wrapper-array/visible.md, and add a links to both pages in docs/en/SUMMARY.

Toilal commented:
Rebased and docs added.

Toilal commented:
Can you restard the build ?

Alex-Sokolov commented on docs/en/api/wrapper/visible.md:
```diff
@@ -0,0 +1,21 @@
+# exists()
```
title missed

Toilal replied:
Oops

Toilal replied:
Fixed

eddyerburgh commented on docs/en/SUMMARY.md:
```diff
@@ -63,6 +64,7 @@
     * [setProps](api/wrapper-array/setProps.md)
     * [trigger](api/wrapper-array/trigger.md)
     * [update](api/wrapper-array/update.md)
+    * [visible](api/wrapper/visible.md)
```
This should be wrapper-array

Toilal replied:
I've made the same as exists() with use the same page.

eddyerburgh replied:
I didn't realize exists uses the same page, this is an error with the exists page.

The wrapper-array pages have different examples that use a wrapper-array

Toilal replied:
Should I fix exists page too ? Or another pull request ?

eddyerburgh commented on docs/en/api/wrapper/visible.md:
```diff
@@ -0,0 +1,21 @@
+Assert `Wrapper` or `WrapperArray` is visible.
+
+Returns false if an ancestor element has `display: none` or `visibility: hidden` style.
+
```
We should add that this is useful for testing whether v-show is working correctly

Toilal replied:
Done

eddyerburgh commented on docs/en/api/wrapper/visible.md:
```diff
@@ -0,0 +1,21 @@
+
+const wrapper = mount(Foo)
+expect(wrapper.visible()).toBe(true)
+expect(wrapper.find('is-not-visible').visible()).toBe(false)
```
is-not-visible should be a class selector, not a tag selector

eddyerburgh commented on src/wrappers/error-wrapper.js:
```diff
@@ -36,6 +36,10 @@ export default class ErrorWrapper implements BaseWrapper {
   }
 
+  visible (): boolean {
+    return false
```
This should throw an error, like the other methods.

Only exists returns false, everything else throws an error to indicate that find was not succesful

Toilal replied:
Don't you think it should match the behavior of `exists()` method ? Having this return an error would make assertions on non-visible elements more verbose.

eddyerburgh replied:
No, I think it should throw an error. `exists` checks if find is successful, so it makes sense to return false. If find('div') is not successful, exists should return false.

Every other method checks something on the Wrapper element/ instance.

We need to make a distinction between a node being visible and a node being present.

eddyerburgh commented on test/unit/specs/mount/Wrapper/visible.spec.js:
```diff
@@ -0,0 +1,135 @@
+    wrapper.update()
+    expect(wrapper.find('.child.ready').visible()).to.equal(true)
+  })
+})
```
Nice tests 👌

Toilal commented:
I just squashed everything and make visible() method fails for non-existent elements. I think it's know OK :p

Toilal committed:
feat(visible): Add visible() method on Wrapper (#327)

eddyerburgh commented on docs/en/api/wrapper/visible.md:
```diff
@@ -0,0 +1,23 @@
+const wrapper = mount(Foo)
+expect(wrapper.visible()).toBe(true)
+expect(wrapper.find('.is-not-visible').visible()).toBe(false)
+expect(wrapper.findAll('div').visible()).toBe(true)
```
Can you remove the findAll examples please

Toilal replied:
Done :)

eddyerburgh commented:
Thanks @Toilal 🙂
